File: sdk/python/tests/release_conformance.py
from __future__ import annotations

import os
import sys
import time

from truyn.client import TruynClient
from truyn.local_node import TruynLocalNodeClient, _request_json, create_envelope

if len(sys.argv) != 2:
    raise SystemExit('relay URL is required')
relay_url = sys.argv[1]
descriptor_url = os.environ.get('TRUYN_CONFORMANCE_DESCRIPTOR_URL')
descriptor_public_key = os.environ.get('TRUYN_CONFORMANCE_DESCRIPTOR_PUBLIC_KEY')
descriptor_identity = os.environ.get('TRUYN_CONFORMANCE_DESCRIPTOR_IDENTITY')
if not descriptor_url or not descriptor_public_key or not descriptor_identity:
    raise SystemExit('descriptor conformance fixture is required')
verified_descriptor = TruynClient(relay_url).fetch_agent_descriptor(descriptor_url, public_key_pem=descriptor_public_key)
assert verified_descriptor['descriptor']['identity'] == descriptor_identity
assert verified_descriptor['selection']['protocol'] == 'TRUYN/1'
assert verified_descriptor['selection']['interface']['type'] == 'https'
assert verified_descriptor['signer']['keyBinding'] == 'identity'

provider = TruynLocalNodeClient.connect(relay_url, name='python-provider')
requester = TruynLocalNodeClient.connect(relay_url, name='python-requester')
capability = f'sdk.release.python.{time.time_ns()}'
provider.offer(capability, {'language': 'python'})
receipt = requester.need(capability, {'question': 'hello'})
need = provider.next_need(timeout_ms=5000)
assert need['needId'] == receipt['needId']
assert need['requester'] == requester.node_id
assert need['capability'] == capability
provider.result(need['needId'], {'ok': True, 'language': 'python'})
result = requester.wait_for_result(receipt['needId'], timeout_ms=5000)
assert result['provider'] == provider.node_id
assert result['verification']['ok'] is True
cancel_receipt = requester.need(capability, {'cancel': True})
revoke = create_envelope(
    type='REVOKE',
    identity=requester.identity,
    payload={'targetId': cancel_receipt['needId'], 'targetKind': 'need', 'reason': 'sdk_conformance_cancel'},
)
_request_json('POST', relay_url.rstrip('/') + '/v1/revoke', body={'envelope': revoke}, headers=requester._auth_headers())
print('PASS python developer-release conformance')

File: sdk/typescript/test/release-conformance.ts
import assert from 'node:assert/strict';
import { TruynClient } from '../src/client.ts';
import { TruynLocalNodeClient } from '../src/local-node.ts';

const relayUrl = process.argv[2];
if (!relayUrl) throw new Error('relay URL is required');
const descriptorUrl = process.env.TRUYN_CONFORMANCE_DESCRIPTOR_URL;
const descriptorPublicKey = process.env.TRUYN_CONFORMANCE_DESCRIPTOR_PUBLIC_KEY;
const descriptorIdentity = process.env.TRUYN_CONFORMANCE_DESCRIPTOR_IDENTITY;
if (!descriptorUrl || !descriptorPublicKey || !descriptorIdentity) throw new Error('descriptor conformance fixture is required');

const descriptorClient = new TruynClient({ relayUrl });
const verifiedDescriptor = await descriptorClient.fetchAgentDescriptor(descriptorUrl, { publicKeyPem: descriptorPublicKey });
assert.equal(verifiedDescriptor.descriptor.identity, descriptorIdentity);
assert.equal(verifiedDescriptor.selection.protocol, 'TRUYN/1');
assert.equal(verifiedDescriptor.selection.interface.type, 'https');
assert.equal(verifiedDescriptor.signer.keyBinding, 'identity');

const provider = await TruynLocalNodeClient.connect({ relayUrl, name: 'typescript-provider' });
const requester = await TruynLocalNodeClient.connect({ relayUrl, name: 'typescript-requester' });
try {
  const capability = `sdk.release.typescript.${Date.now()}.${Math.random()}`;
  await provider.offer(capability, { language: 'typescript' });
  const receipt = await requester.need(capability, { question: 'hello' });
  const need = await provider.nextNeed({ timeoutMs: 5_000 });
  assert.equal(need.needId, receipt.needId);
  assert.equal(need.requester, requester.nodeId);
  assert.equal(need.capability, capability);
  await provider.result(need.needId, { ok: true, language: 'typescript' });
  const result = await requester.waitForResult(receipt.needId, { timeoutMs: 5_000 });
  assert.equal(result.provider, provider.nodeId);
  assert.equal(result.verification.ok, true);
  const cancelReceipt = await requester.need(capability, { cancel: true });
  await requester.runtime.revoke(cancelReceipt.needId, 'sdk_conformance_cancel', { targetKind: 'need' });
  process.stdout.write('PASS typescript developer-release conformance\n');
} finally {
  provider.close();
  requester.close();
}
